A few fixes to ensure smooth MCNP6 ACE library conversion

paulromano opened the pull request:
This pull request makes a few changes that I found to be necessary in order to convert MCNP6 cross section libraries.

- Add a few names to our `_THERMAL_NAMES` dictionary so that none are unrecognized.
- Handle continuation lines in xsdir files
- When writing HDF5 files, open it in 'w' mode (overwrite, rather than append which is default). I found this to be necessary for a few of the thermal tables because in the MCNP6 data, they actually have some tables repeated with different names, e.g. h-zr.20t and h/zr.20t appear in the same file with exactly the same data.

smharper commented:
Trying to test this out but I need to clear off some disk space.  My poor hard drive isn't a fan of this data format development!

smharper commented:
I tried this out with `xsdir_mcnp5_endfb-7.1` and got this error:
```
Converting 2004.62c (ACE) to He4.62c (HDF5)
Traceback (most recent call last):
  File "/home/smharper/openmc/scripts/openmc-ace-to-hdf5", line 120, in <module>
    table, args.metastable)
  File "/home/smharper/openmc/openmc/data/neutron.py", line 431, in from_ace
    rx = Reaction.from_ace(ace, i)
  File "/home/smharper/openmc/openmc/data/reaction.py", line 493, in from_ace
    ace, ace.jxs[11], lnw, rx))
  File "/home/smharper/openmc/openmc/data/angle_energy.py", line 106, in from_ace
    "distribution law {}".format(law))
ValueError: Unsupported ACE secondary energy distribution law 67
```

I'll try it with xsdir_mcnp6.1 next

nelsonag commented:
I have MCNP5 endf70 data on this machine right now. So I'll report back on how this works with that.

nelsonag commented on scripts/openmc-ace-to-hdf5:
```diff
@@ -111,25 +118,25 @@ for filename in ace_libraries:
             # Continuous-energy neutron data
```
@paulromano So I found an issue related more to runtime/workload than correctness.
As you know, ENDF-VII.0 which ships with MCNP has files endf70a up to endf70k and endf70sab.  endf70a, for example, contains data for many isotopes (up to V-Nat) with 5 temperatures for each.  the xsdir or cross_sections.xml file will have a separate entry for every one of those iso/temp combinations, all with the filename of endf70a.
The way this is written the script will see the 250-ish something tables within endf70a, and process each and everyone of them 250-ish times.

nelsonag replied:
You _could_ fix this by just removing duplicate etnries from ace_libraries.  To approximate this, I simplified my cross_sections.xml file so that I only had one table from each library file; seems to work.  

The problem with this is that if someone took care to remove data they didnt need from their xsdir/xsdata/cross_sections.xml file (like non-relevant temperatures or isotopes), then this strategy would over-rule that explicit decision the user has made.

paulromano replied:
It looks like this was only really a problem for cross_sections.xml (I'll get that fixed). For xsdir and xsdata, the script was already checking for duplicate paths.

It would take quite a bit more work to actually be able to pick out specific tables rather than just specify files, so I'm going to leave this with the simple solution (removing duplicate entries).

nelsonag commented:
One more thing, since this seems to be the right place to fix it right now: Jeff 3.1.2 fails for me when it accesses line 100 of openmc/data/function.py.  ('contined' should be 'contained')

smharper commented:
The `xsdir_mcnp6.1_endfb-7.1` references the same ACE files so I still get the
```
ValueError: Unsupported ACE secondary energy distribution law 67
```

paulromano commented:
@smharper I've addressed your comments and made a few little fixes so that converting based on the MCNP-distributed xsdir file should work. It will now just show an error message if converting a particular table fails. Give it another whirl and let me know if you run into any problems.

nelsonag commented:
Nothing more from me, thanks @paulromano!

smharper commented:
`IncidentNeutron` tries to figure out the library suffix from the ACE file name, but these are different than what MCNP uses in the xsdir file.  E.g. 1001.80c in xsdir points to the 1001.710nc ACE file so OpenMC thinks the name.  Should be pretty straightforward to get the proper library name in `openmc-ace-to-hdf5` and pass it as an argument to `IncidentNeutron.from_ace`

paulromano commented:
@smharper `IncidentNeutron` uses the name from `ace.Table`. The issue with the MCNP6 files is that they are ACE 2.0.0 files. If you look at the first line of any of those, such as 1001.710nc, you'll see that the 710nc is there in the file, but a few lines down there is also 80c. This was done for compatibility reasons -- older versions of MCNP don't know about the new header, so to get around this, they include an old-style header in the comment block, and then the xsdir file indicates that the table starts at line 4; effectively the first three lines of the file aren't read. Right now, we do read the ACE 2.0.0 header and get the name from the first line, which is 1001.710nc. We could change it to look further down and get the "old-style" 80c suffix and use that instead if there is a compelling reason.

smharper commented:
Ah, I see.  We'll I don't think we need to support legacy MCNP formats so I'm happy with the PR as-is.  Merging!
